### ISSUE TITLE
torch.topk on XPU returns invalid (-1) indices with NaN input at batch size >= ~512

### ISSUE THREAD

**Issue by mganczarenko** [intel/torch-xpu-ops#4839]
## Summary

`torch.topk` on XPU returns out-of-range indices (`-1`) when the input contains a NaN row and the batch dimension is >= ~512. CPU on the identical seed/input returns valid indices. This is not an alternate/valid NaN-ordering convention — `-1` is never a valid `topk` index, and downstream `gather()` on that index aborts the process via a device-side bounds assert rather than raising a catchable exception.

Likely related to intel/torch-xpu-ops#3777 ("[Triton] torch.topk returns wrong values on XPU", closed without a linked fix) — same op, overlapping version window — but that repro is `[8,8]` float32 with no NaN, so we haven't confirmed it's the identical underlying defect.

## Repro

```python
import torch

torch.manual_seed(0)
go_cpu = torch.randn(512, 32, dtype=torch.bfloat16)
go_cpu[0] = float("nan")

values_cpu, ids_cpu = torch.topk(go_cpu, k=4, dim=-1, sorted=False)
print("CPU row0 ids:   ", ids_cpu[0].tolist())

go_xpu = go_cpu.to("xpu")
values_xpu, ids_xpu = torch.topk(go_xpu, k=4, dim=-1, sorted=False)
print("XPU row0 ids:   ", ids_xpu[0].tolist())

bad = (ids_xpu < 0) | (ids_xpu >= 32)
print("XPU bad_count:  ", bad.sum().item())

try:
    go_cpu.gather(1, ids_xpu.cpu())
    print("CPU gather with XPU's ids: succeeded (unexpected)")
except (IndexError, RuntimeError) as e:
    print("CPU gather with XPU's ids: raised", type(e).__name__, "->", e)

# On XPU, the same gather aborts the process with a device_assert
# (SIGABRT) instead of raising a catchable exception.
go_xpu.gather(1, ids_xpu)
```

## Observed output

```
CPU row0 ids:    [22, 21, 20, 23]
XPU row0 ids:    [0, -1, -1, -1]
XPU bad_count:   3
CPU gather with XPU's ids: raised RuntimeError -> index -1 is out of bounds for dimension 1 with size 32
AssertHandler::printMessage
.../torch-xpu-ops/src/ATen/native/xpu/sycl/ScatterGatherKernels.cpp:233: operator(): global id: [1,0,0], local id: [1,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size_ && "scatter gather kernel index out of bounds"` failed.
```
Process exits with SIGABRT (134).

## Trigger conditions (from further isolation, happy to share the sweep script)

- Requires a NaN row in the input — fully finite inputs (including degenerate all-zero/all-equal ties, and +-Inf) never reproduce, at any batch size tested.
- Requires batch size >= ~512 — fails at 512/1024/2048/4096/8192, passes at 16/64/128/256.
- Reproduces with `sorted=False`; not yet swept for `sorted=True`.

## Environment

- `torch==2.13.0+xpu`
- `triton-xpu==3.7.2`
- Intel Arc Pro B70 (30.3GiB), driver `intel-opencl-icd 26.18.38308.1`
- Does **not** reproduce on `torch==2.12.0+xpu` (older stack, otherwise same environment) — bisected by dependency bump alone, no application-level code changed across the bisect.

## Downstream impact

This corrupts MoE expert routing in vLLM (`grouped_topk()`, used by DeepSeek-V2/V3-style and LiquidAI LFM2 gates) when the router's gating output happens to contain a NaN — the corrupted index is then consumed by a fused Triton kernel doing `gather`, which asserts one step later than the plain-`gather` repro above shows, but from the same root cause. We understand from internal driver-team follow-up that a fix (a loop-strength-reduction wrap-around bug in IGC's shader codegen) has landed internally but is not yet in a public/GA driver release. We have not been able to test our exact repro (bf16, NaN-triggered) against the fixed driver build to confirm it resolves this trigger.

cc @astachowiczhabana

**Comment by mganczarenko** [comment]
Internal follow-up on our end: driver/VAL team traced this to a loop-strength-reduction wrap-around bug in IGC codegen, hit by `SegmentedGroupRadixSelectPairsFunctor`/`SegmentedGroupRadixSortPairsFunctor`. Fix has landed internally (Intel GSD-12849, status: Implemented) but is not yet in a public/GA driver release as of this writing. We have not been able to test our exact repro (bf16, NaN-triggered) against the fixed driver build to confirm it resolves this specific trigger.

**Comment by astachowiczhabana** [comment]
**Root cause**

Short version: the kernel that computes topk keeps some "empty" slots in a scratch buffer and marks them with the value -infinity and the index -1. It then decides which entries to keep by comparing only the values. Both NaN and a real -infinity input fail that comparison, so the empty slots are never replaced, and the placeholder index -1 gets written out as a real result.

**Some background first**

torch.topk(x, k) returns the k largest values in each row, and the position (index) of each one. Those indices are the part that matters here, because whatever consumes them will use them to look values up -- for example x.gather(1, indices) in the report.

XPU has three different implementations of topk and picks one at runtime depending on the shape (see TensorTopKKernel.cpp and TensorTopKSbtopkKernel.cpp). The one in this bug is the fastest and most specialised: the "sub-group" kernel, chosen only when k <= 8 and there are many rows. It is in TensorTopKSbtopkKernelImpl.h.

It works like this. A "sub-group" is a batch of 32 work items (lanes) that run in lockstep on the GPU. One sub-group handles one row of your tensor:

The row is split across the 32 lanes. Each lane walks its own chunk and keeps a small sorted list of the best K values it has seen, together with their indices.
The lanes then combine their lists in pairs -- lane 0 with lane 1, then 0 with 2, and so on -- until lane 0 holds the best K for the whole row. That merging step is bitonic_merge().
Lane 0 writes the K values and their indices out.
Where it goes wrong

At the start, every lane's list of K entries is filled with a placeholder, because nothing has been seen yet:

value = -infinity   (or the smallest possible integer, for integer types)
index = -1

The idea is that any real input beats a placeholder and pushes it out. The merge step decides that with a plain comparison on the value alone -- roughly if (their_value > my_value) take theirs.

That is the bug, and it fails in two separate ways:

NaN. By definition, every comparison involving NaN is false. NaN > x is false, NaN < x is false, even NaN == NaN is false. So NaN > -infinity is false, and a NaN never displaces a placeholder.

A real -infinity (or INT_MIN). The placeholder value is not a magic value -- it is a perfectly legal input value. And -infinity > -infinity is also false, because the comparison is strict. So a genuine -infinity in your data cannot displace a placeholder either. No NaN is needed to trigger this bug at all.

When a placeholder survives to the end, lane 0 writes out its index, -1. Now torch.topk has returned -1 as a position in your row. The gather() that follows reads out of bounds, and because a GPU cannot raise a Python exception from inside a kernel, you get the device-side assert / abort in the report instead of a normal error.

Whether you see it depends on how many placeholders are left over, which is why narrow rows fail and wide rows appear fine. Each lane loads several elements at a time (1, 2, 4 or 8, chosen from the dtype and row width), so with a 32-element row and k=4 each lane only ever sees one element and 3 of its 4 slots stay empty. That matches the reported output exactly: indices [0, -1, -1, -1] at 32 columns, and [0, 1, -1, -1] at 64 columns.

**The telling detail**

The other function in the same file, insert(), already knew about this. It tracked how many slots were filled in a count variable and used it to guard the comparison, and its comment even spells out the all--infinity case:

// ... an input value equal to the sentinel (e.g. all -inf for
// largest=true) would always stop at position K-1 ...

bitonic_merge() never tracked a count and had no equivalent guard. So the buffer was protected while a single lane was filling it, and unprotected the moment two lanes merged. That asymmetry is the whole bug.

The fix is to compare (value, index) pairs rather than values alone: an entry whose index is negative is empty and must lose to everything, and NaN has to be ordered explicitly (PyTorch treats NaN as larger than every real number, so it comes first for largest=True and is excluded for largest=False). Then the same comparison can be used by both insert() and bitonic_merge(), so they cannot disagree again.

**The impact is wider than the report**

Invalid -1 indices, then an abort. The reported symptom.
Silently wrong values, with no crash. When a NaN fails to displace anything, it is dropped from the result and a smaller real value is promoted in its place. This affects every shape that reaches this kernel, including the wide rows where all the indices look valid. No crash, no warning, just a wrong answer -- arguably worse than the abort.
No NaN required. A row of all -infinity (largest=True), all +infinity (largest=False), or all INT_MIN (int32) reproduces the -1 indices.
Two details in the report worth correcting

This is not an IGC codegen bug. It reproduces on a newer driver (26.22.38646.6) and is fully explained by the SYCL source.
"batch >= 512" is not portable. The real trigger is the threshold that selects this kernel: nsegments >= EU_count * HW_threads_per_EU / 4 in TensorTopKSbtopkKernel.cpp. That is 320 rows on the GPU I tested (160 EUs) and 512 on the reporter's, which is why 512 looked like the magic number. sorted=False and bfloat16 are also incidental -- sorted=True and fp16/fp32/fp64 fail identically.

**How it got here (commit history)**

8eaa591f  2026-04-29  Add subgroup topk kernel for XPU (part1 of #3369) (#3371)
5872efe1  2026-05-19  Revert "Add subgroup topk kernel ... (#3371)" (#3707)
9054dacf  2026-05-25  Split subgroup topk kernel per-K for parallel AOT compilation (#3683)
6336b4f8  2026-05-26  Add single workgroup topk kernel for XPU (part 2 of #3369) (#3372)
d27392a0  2026-06-15  Use C++20 bit_ceil/bit_floor instead of hand-rolled helpers (#3950)
ffdf468e  2026-08-12  Fix TopK AOT (#4907)

The bug arrived with the kernel itself in #3371 and has never been touched since. Reading the list in order:

#3371 introduced the sub-group kernel, including the value-only comparison in bitonic_merge(). This is the commit that introduced the bug.
#3707 reverted #3371 -- but for an unrelated reason (a build timeout in stock CD), not for this bug.
#3683 re-landed it. The comparison logic is byte-for-byte the same; the only change was splitting the code into per-K files so AOT compilation could run in parallel. So the revert-and-reland did not touch the defect.
#3372 is the interesting one. It added the other new topk kernel (the single-workgroup radix one) and, in passing, "fixes NaN handling in SortingRadixSelect.h TopKTypeConfig::convert" -- mapping NaN to the maximum radix value so it sorts correctly there. So NaN in topk was a known concern one day after #3683 re-landed the sub-group kernel, but only the radix path was fixed and the sub-group kernel was never revisited.
#3950 and #4907 are unrelated cleanups (a C++20 helper swap, and moving kernel properties onto the functor to avoid a JIT recompile).
Since these are all recent, this affects only builds that include #3371 / #3683; anything older routes through the fallback path and is unaffected.

**Note on cost**

Making the comparison correct is not free -- it adds an index check, and a NaN check for floating-point types, to the innermost loop. Measured on one device across 25 shapes that use this kernel, it costs roughly 11% more time per call (geomean; a few shapes get slightly faster, the worst is about 30% slower). That is still well ahead of the fallback path this kernel replaced, so fixing looks clearly better than reverting, but it is worth knowing before review.
